### ISSUE TITLE
Digest data and size for Fapi_Sign API

### ISSUE THREAD

**Issue by dwaraka1982** [tpm2-software/tpm2-tss#2226]
We see that Fapi_Sign takes digest of fixed sizes mostly 20, 32 bytes etc and then accordingly selects the Hash Algo(SHA1, SHA256 ...).

Is there any Fapi/Esys API to generate this digest given a data of any random size ? 

We see there is one API "Esys_Hash" but again that is expecting the data to be in fixed size and it will generate a digest out of it. What we are looking for is for some API which can generate the digest given any random data and then feed the digest to the Fapi_Sign for signing and subsequent verification.

**Comment by williamcroberts** [comment]
TL;DR Use OpenSSL to hash

IIUC, the only reason to have the TPM hash your code is because you need to use a restricted key. To use a restricted key, you need the validation ticket from the TPM Hash call to pass to TPM Sign. So really, you should just hash it using whatever standard crypto library is in use for your environment. Since you're in C, you would probably want to use something like OpenSSL:
  - https://wiki.openssl.org/index.php/EVP_Message_Digests

This will produce the digest you can pass to Fapi_Sign.

ESYS has support for multiple hash calls, you can use a Sequence:
  - Esys_HashSequenceStart
  - Esys_SequenceUpdate
  - Esys_SequenceComplete

**Comment by dwaraka1982** [comment]
Ok got it, thanks for the inputs, i will close this issue now
